casa/IO/MultiFile.cc

```diff
@@ -36,191 +37,752 @@
       }
     }
+    if (nfree != Int64(itsFreeBlocks.size())) {
+      itsFreeBlocks.resize (nfree);
+    }
   }
 
   void MultiFile::readBlock (MultiFileInfo& info, Int64 blknr,
                              void* buffer)
   {
-    itsIO.pread (itsBlockSize, info.blockNrs[blknr] * itsBlockSize, buffer);
+    if (! itsTestMode) {
+      itsIO->pread (itsBlockSize, info.blockNrs[blknr] * itsBlockSize, buffer);
+      if (itsUseCRC) {
+        checkCRC (buffer, info.blockNrs[blknr]);
+      }
+    }
   }
 
   void MultiFile::writeBlock (MultiFileInfo& info, Int64 blknr,
                               const void* buffer)
   {
-    itsIO.pwrite (itsBlockSize, info.blockNrs[blknr] * itsBlockSize, buffer);
+    if (! itsTestMode) {
+      itsIO->pwrite (itsBlockSize, info.blockNrs[blknr] * itsBlockSize, buffer);
+    }
+    if (itsUseCRC) {
+      storeCRC (buffer, info.blockNrs[blknr]);
+    }
+  }
+
+  void MultiFile::storeCRC (const void* buffer, Int64 blknr)
+  {
+    if (blknr >= Int64(itsCRC.size())) {
+      itsCRC.resize (blknr+1);
+    }
+    itsCRC[blknr] = calcCRC (buffer, itsBlockSize);
+  }
+
+  void MultiFile::checkCRC (const void* buffer, Int64 blknr) const
+  {
+    AlwaysAssert (blknr < Int64(itsCRC.size()), AipsError);
+    uInt crc = calcCRC (buffer, itsBlockSize);
+    if (crc != itsCRC[blknr]) {
+      throw AipsError ("Mismatch in CRC of MultiFile block " +
+                       String::toString(blknr));
+    }
+  }
+
+  uInt MultiFile::calcCRC (const void* buffer, Int64 size) const
+  {
+    // The algorithm is taken from crctester.c by Sven Reifegerste
+    // (www.zorc/reflex)
+    // See also https://www.lammertbies.nl/comm/info/crc-calculation
+    const uChar* buf = static_cast<const uChar*>(buffer);
+    const uInt crcinit = 0x46af6449;
+    //# The above value is calculated from:
+    //#         const uInt highbit = (uInt)1<<(32-1);
+    //#		crcinit = 0xffffffff;
+    //#		for (i=0; i<32; i++) {
+    //#			bit = crcinit & 1;
+    //#			if (bit) crcinit^= polynom;
+    //#			crcinit >>= 1;
+    //#			if (bit) crcinit|= highbit;
+    //#		}	
+    // Normal lookup table algorithm with augmented zero bytes.
+    uInt crc = crcinit;
+    for (Int64 i=0; i<size; ++i) {
+      crc = ((crc << 8) | buf[i]) ^ itsCRCTab[ (crc >> (32-8))  & 0xff];
+    }
+    // Augment zero bytes.
+    for (size_t i=0; i<32/8; ++i) {
+      crc = (crc << 8) ^ itsCRCTab[ (crc >> (32-8))  & 0xff];
+    }
+    crc^= 0xffffffff;
+    return crc;
+  }
+
+  void MultiFile::fillCRCTable()
+  {
+    // CRC-32 parameters
+    // 'polynom' is the standard CRC-32 polynom without leading '1' bit
+    const uInt polynom = 0x4c11db7;
+    const uInt highbit = (uInt)1<<(32-1);
+    // make CRC lookup table used by table algorithms
+    for (int i=0; i<256; i++) {
+      uInt crc = i;
+      crc<<= 32-8;
+      for (int j=0; j<8; j++) {
+        uInt bit = crc & highbit;
+        crc<<= 1;
+        if (bit) crc^= polynom;
+      }			
+      itsCRCTab[i]= crc;
```

Comment:
I assume this CRC table is identical for every object or not?
I would recommend to calculate this table compiletime, but it seems Casacore only requires C++11.

Reply:
It's indeed the same for all. I could make it a singleton which requires a threading guard.
How can you calculate it compile-time?

Reply:
Using C++14 or preferably C++17 ;-)
There is no need for a singleton to do this in C++11.

Reply:
A function like this at the top of the file should do the trick and is thread-safe and guards against the static initialization fiasco. (I didn't compile this.)
```
static const std::array<uInt, 256>& CRCTable()
{
  static std::array<uInt, 256> result = [] {
    std::array<uInt, 256> result ;

    const uInt polynom = 0x4c11db7;
    const uInt highbit = (uInt)1<<(32-1);
    // make CRC lookup table used by table algorithms
    for (int i=0; i<256; i++) {
      uInt crc = i;
      crc<<= 32-8;
      for (int j=0; j<8; j++) {
        uInt bit = crc & highbit;
        crc<<= 1;
        if (bit) crc^= polynom;
      }			
      result[i]= crc;    
    }
  }();
  return result;
}